src/dockerfile-from-checker/src/main.go

```diff
@@ -0,0 +1,186 @@
+package main
+
+import (
+	"fmt"
+	"io/fs"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/moby/buildkit/frontend/dockerfile/instructions"
+	"github.com/moby/buildkit/frontend/dockerfile/parser"
+	"github.com/spf13/cobra"
+)
+
+var rootCmd = &cobra.Command{}
+
+func init() {
+	rootCmd = &cobra.Command{
+		Use:  "dockerfile-checker <dir>",
+		Args: cobra.ExactArgs(1),
+		Run:  rootFunc,
+	}
+}
+
+func rootFunc(cmd *cobra.Command, args []string) {
+	var paths []string
+	ignorePaths := make(map[string]struct{})
+
+	ignoreRelPaths, err := rootCmd.Flags().GetStringSlice("ignore-files")
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, ign := range ignoreRelPaths {
+		absPath, err := filepath.Abs(ign)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		ignorePaths[absPath] = struct{}{}
+	}
+
+	filepath.Walk(args[0], func(p string, info fs.FileInfo, err error) error {
+		if info.Name() == "vendor" {
+			return filepath.SkipDir
+		}
+		if info.IsDir() {
+			return nil
+		}
+
+		if info.Name() == "Dockerfile" {
+			absPath, err := filepath.Abs(p)
+			if err != nil {
+				log.Fatal(err)
+			}
+			_, ok := ignorePaths[absPath]
+			if !ok {
+				// path is not on ignore list
+				paths = append(paths, absPath)
+			}
+		}
+		return nil
+	})
+
+	checkDockerfiles(paths)
+}
+
+func checkDockerfiles(paths []string) {
+	var f *os.File
+	var err error
+
+	froms2dockerfile := make(map[string][]string)
+	for _, filename := range paths {
+		f, err = os.Open(filename)
+		if err != nil {
+			log.Fatalf("could not open %s: %+v", filename, err)
+		}
+		defer f.Close()
+
+		froms := parseDockerfile(f)
+		for _, from := range froms {
+			fns := append(froms2dockerfile[from], filename)
+			froms2dockerfile[from] = fns
+		}
+	}
+
+	checkInconsistencies(froms2dockerfile)
+}
+
+func main() {
+	rootCmd.Flags().StringSliceP("ignore-files", "i", []string{}, "ignores specified Dockerfile; can be used several times")
+
+	err := rootCmd.Execute()
+	if err != nil {
+		os.Exit(1)
+	}
+}
+
+func checkInconsistencies(froms2dockerfile map[string][]string) {
+	type tagFile struct {
+		tag      string
+		file     string
+		fullname string
+	}
+
+	image2TagFile := make(map[string][]tagFile)
+	for from, files := range froms2dockerfile {
+		for _, file := range files {
+			splits := strings.Split(from, ":")
+			if len(splits) < 1 {
+				continue
+			}
+			tag := splits[len(splits)-1]
+			image := strings.Join(splits[:len(splits)-1], "")
+			if image == "" {
+				continue
+			}
+			tf := tagFile{
+				tag:      tag,
+				file:     file,
+				fullname: from,
+			}
+			image2TagFile[image] = append(image2TagFile[image], tf)
+		}
+	}
+
+	for _, tfs := range image2TagFile {
+		for i := 1; i < len(tfs); i++ {
+			tf := tfs[i]
+			if tf.tag != tfs[i-1].tag {
+				fmt.Printf("tags differ for image %s in files %s and %s\n", tf.fullname, tf.file, tfs[i-1].file)
+				os.Exit(1)
```

Comment:
Stylistically it would make more sense to move the exit to the main function. And I think exit(1) is normally for input errors (unknown option, input file does not exist).